doc/lsst.daf.butler/exprParser.rst

```diff
@@ -0,0 +1,176 @@
+    encoding chosen
+
+Numbers
+    Integer numbers are series of decimal numbers optionally preceded by
+    minus sign. Parser does not support octal/hexadecimal numbers. Floating
+    point numbers use standard notation with decimal point and/or exponent.
+    For numbers parser passes a string representation of a number to
+    downstream registry code to avoid possible rounding issues.
+
+Range literals
+    This sort of literals is allowed inside ``IN`` expression only. It consists
+    of two integer literals separated by double dots and optionally followed by
+    a colon and one more integer literal. Two integers define start and stop
+    values for the range, both are inclusive values. Optional third integer
+    defines stride value which defaults to 1, it cannot be negative. Ranges
+    are equivalent to a sequence of integers (but not to intervals of floats).
+
+Examples of range literals:
+
+* ``1..5`` -- equivalent to ``1,2,3,4,5``
+* ``1..10:3`` -- equivalent to ``1,4,7,10``
+* ``-10..-1:2`` -- equivalent to ``-10,-8,-6,-4,-2``
+
+Identifiers
+-----------
+
+Identifiers represent values external to a parser such as values stored in a
+database. Parser itself cannot define identifiers or their values, it is
+responsibility of translation layer (registry) to map identifiers into
+something sensible. Like in most programming languages identifier starts with
+a letter or underscore followed by zero or more letters, underscores, or
+digits. Parser also supports dotted identifiers consisting of two simple
+identifiers separated by a dot. Identifiers are case-sensitive on parser side
+but individual database back-ends may have special rules about case
+sensitivity.
+
+In current implementation simple identifiers are used by registry to represent
+dimensions, e.g. ``visit`` identifier is used to represent a value of
+``visit`` dimension in registry database. Dotted identifiers are mapped to
+tables and columns in registry database, e.g. ``detector.raft`` can be used
+for accessing raft name (obviously dotted names need knowledge of database
+schema and how SQL query is built).
+
+Unary arithmetic operators
+--------------------------
+
+Two unary operators ``+`` (plus) and ``-`` (minus) can be used in the
+expressions in front of (numeric) literals, identifiers, or other
+expressions which should evaluate to a numeric value.
+
+Binary arithmetic operators
+---------------------------
+
+Language supports five arithmetic operators - ``+`` (add), ``-`` (subtract),
+``*`` (multiply), ``/`` (divide), and ``%`` (modulo). Usual precedence rules
+apply to these operators. Operands for them can be anything that evaluates to
+a numeric value.
+
+Comparison operators
+--------------------
+
+Language supports set of regular comparison operators - ``=``, ``!=``, ``<``,
+``<=``, ``>``, ``>=``. This can be used on operands that evaluate to a numeric
+values, for (in)equality operators operands can also be boolean expressions.
```

Comment:
Maybe add a note that equality comparison is a single `=`, like SQL but unlike Python?

Reply:
I added this:

    .. note :: The equality comparison operator is a single ``=`` like in SQL, not
        double ``==`` like in Python or C++.